packages/runtime/container-runtime/src/dataStore.ts

```diff
@@ -0,0 +1,151 @@
+
+/**
+ * Type guard that returns true if the given alias message is actually an instance of
+ * a class which implements @see IDataStoreAliasMessage
+ * @param maybeDataStoreAliasMessage - message object to be validated
+ * @returns True if the @see IDataStoreAliasMessage is fully implemented, false otherwise
+ */
+ export const isDataStoreAliasMessage = (
+    maybeDataStoreAliasMessage: any,
+): maybeDataStoreAliasMessage is IDataStoreAliasMessage => {
+    return typeof maybeDataStoreAliasMessage?.internalId === "string"
+        && typeof maybeDataStoreAliasMessage?.alias === "string";
+};
+
+/**
+ * A fluid router with the capability of being assigned an alias
+ */
+ export interface IDataStore extends IFluidRouter {
+    /**
+     * Attempt to assign an alias to the datastore.
+     * If the operation succeeds, the datastore can be referenced
+     * by the supplied alias.
+     *
+     * @param alias - Given alias for this datastore.
+     */
+    trySetAlias(alias: string): Promise<boolean>;
+}
+
+export const channelToDataStore = (
+    fluidDataStoreChannel: IFluidDataStoreChannel,
+    internalId: string,
+    runtime: ContainerRuntime,
+    logger: ITelemetryLogger,
+): IDataStore => new DataStore(fluidDataStoreChannel, internalId, runtime, logger);
+
+enum AliasState {
+    Aliased = "Aliased",
+    Aliasing = "Aliasing",
+    None = "None",
+}
+
+class DataStore implements IDataStore {
+    private aliasState: AliasState = AliasState.None;
+    private alias: string | undefined;
+
+    async trySetAlias(alias: string): Promise<boolean> {
+        assert(this.runtime.attachState === AttachState.Attached, "Trying to alias when not attached!");
+        switch (this.aliasState) {
+            // If we're already aliasing, fail the current function call
+            case AliasState.Aliasing: return false;
+            // If this datastore is already aliased, return true only if this
+            // is a repeated call for the same alias
+            case AliasState.Aliased: return this.alias === alias;
+            // There is no current or past alias operation for this datastore,
+            // it is safe to continue execution
+            case AliasState.None: break;
+            default: unreachableCase(this.aliasState);
+        }
+
+        this.aliasState = AliasState.Aliasing;
+        this.fluidDataStoreChannel.bindToContext();
+
+        const message: IDataStoreAliasMessage = {
+            internalId: this.internalId,
+            alias,
+        };
+
+        return this.ackBasedPromise<boolean>((resolve) => {
+            this.runtime.submitDataStoreAliasOp(message, resolve);
+        }).then((succeeded) => {
+            if (succeeded) {
+                this.aliasState = AliasState.Aliased;
+                this.alias = alias;
+            } else {
+                this.aliasState = AliasState.None;
+            }
+
+            return succeeded;
+        }).catch((error) => {
+            this.logger.sendErrorEvent({
```

Comment:
If this exception is not expected, then maybe use fail-fast approach and simply close container with this error?
This is only valid if this is indeed some kind of exceptional case that we can't foresee.
I like smaller states and processes to reason about (including analyzing errors), and not leaving container in unknown states.

Reply:
That said, dispose is one such case. But I'd argue you want to somehow figure out a solution not to get here on dispose, as we do not want to see these error events in telemetry and keeps saying "you can ignore them". Error table should have only actionably (true) errors.

Reply:
I'd rather have this running and see if we actually hit it. It's hard to tell right now how this will go in the wild and closing down the container seems a bit too aggressive. If they are rare, I'd go with closing it down. If they're noisy, we can remove them.